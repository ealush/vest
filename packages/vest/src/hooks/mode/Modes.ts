export enum Modes {
  ALL,
  EAGER,
}
