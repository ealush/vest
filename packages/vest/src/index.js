import any from "anyone/any";
import enforce from "n4s/dist/enforce.min";
import { VERSION } from "./constants";
import test from "./core/test";
import validate from "./core/validate";
import { draft, only, skip, warn } from "./hooks";
import { singleton, runWithContext } from "./lib";

export default singleton.register({
  Enforce: enforce.Enforce,
  VERSION,
  any,
  draft,
  enforce,
  only,
  runWithContext,
  skip,
  test,
  validate,
  warn,
});
