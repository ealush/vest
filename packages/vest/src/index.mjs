import enforce from '../../n4s/src/enforce';
import create from './core/suite/create';
import test from './core/test';
import { draft, only, skip, warn, group } from './hooks';

const VERSION = VEST_VERSION;

export default {
  VERSION,
  create,
  draft,
  enforce,
  group,
  only,
  skip,
  test,
  warn,
};

export { VERSION, create, draft, enforce, group, only, skip, test, warn };
