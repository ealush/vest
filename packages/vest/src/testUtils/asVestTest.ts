import { invariant } from 'vest-utils';

import { ErrorStrings } from 'ErrorStrings';
import { TIsolateTest } from 'IsolateTest';
import { isIsolateTest } from 'isIsolateTest';

export function asVestTest(value: any): TIsolateTest {
  invariant(isIsolateTest(value), ErrorStrings.EXPECTED_VEST_TEST);
  return value;
}
