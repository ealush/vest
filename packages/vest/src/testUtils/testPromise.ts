export function TestPromise(cb: (done: () => void) => void): Promise<void> {
  return new Promise<void>(done => cb(done));
}
