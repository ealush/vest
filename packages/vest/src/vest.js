import create from 'createSuite';
import enforce from 'enforce';
import { only, skip, warn, group } from 'hooks';
import test from 'test';

const VERSION = __LIB_VERSION__;

export default {
  VERSION,
  create,
  enforce,
  group,
  only,
  skip,
  test,
  warn,
};
