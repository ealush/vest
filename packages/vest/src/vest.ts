import { enforce } from 'n4s';

import create from 'create';
import context from 'ctx';
import each from 'each';
import { only, skip } from 'exclusive';
import group from 'group';
import include from 'include';
import { eager } from 'mode';
import omitWhen from 'omitWhen';
import optional from 'optionalTests';
import skipWhen from 'skipWhen';
import { test } from 'test';
import warn from 'warn';

const VERSION = __LIB_VERSION__;
export {
  test,
  create,
  each,
  only,
  skip,
  warn,
  group,
  optional,
  skipWhen,
  omitWhen,
  enforce,
  VERSION,
  context,
  include,
  eager,
};
