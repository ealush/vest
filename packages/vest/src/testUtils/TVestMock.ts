import { TFieldName, TGroupName } from 'SuiteResultTypes';
import * as vest from 'vest';

export type TVestMock = typeof vest;

export type TTestSuiteCallback = (...args: any[]) => void;
export type TTestSuite = vest.Suite<TFieldName, TGroupName, TTestSuiteCallback>;
