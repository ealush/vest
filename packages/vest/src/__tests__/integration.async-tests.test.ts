import { TTestSuite } from 'testUtils/TVestMock';
import wait from 'wait';

import { TestPromise } from '../testUtils/testPromise';

import { Modes } from 'Modes';
import * as vest from 'vest';

function genSuite() {
  return vest.create(() => {
    vest.mode(Modes.ALL);
    vest.test('field_1', 'field_statement_1', () => false);
    vest.test('field_2', 'field_statement_2', () => {
      vest.enforce(2).equals(3);
    });
    vest.test('field_3', 'field_statement_3', jest.fn());
    vest.test('field_4', 'field_statement_4', () => {
      vest.warn();
      throw new Error();
    });
    vest.test('field_4', 'field_statement_4', () => {
      vest.warn();
    });
    vest.test('field_5', 'field_statement_5', () => false);
    vest.test('field_5', 'field_statement_6', () => false);
    vest.test('field_6', 'async_statement_1', async () => {
      await wait(250);
    });
    vest.test('field_7', () => Promise.reject('async_statement_2'));
  });
}

let suite: TTestSuite;
describe('Stateful behavior', () => {
  let result,
    callback_1 = jest.fn(),
    callback_2 = jest.fn(),
    callback_3 = jest.fn(),
    control = jest.fn();

  beforeEach(() => {
    suite = genSuite();
  });

  beforeAll(() => {
    callback_1 = jest.fn();
    callback_2 = jest.fn();
    callback_3 = jest.fn();
    control = jest.fn();
  });

  test('Should have all fields', () =>
    TestPromise(done => {
      // ❗️Why is this test async? Because of the `resetState` beforeEach.
      // We must not clean up before the suite is actually done.
      result = suite().done(done);
      expect(result.tests).toHaveProperty('field_1');
      expect(result.tests).toHaveProperty('field_2');
      expect(result.tests).toHaveProperty('field_4');
      expect(result.tests).toHaveProperty('field_5');
      expect(result.tests).toHaveProperty('field_6');
      expect(result.tests).toHaveProperty('field_7');
      expect(result.hasErrors('field_7')).toBe(false);
      expect(result.tests).toMatchSnapshot();
    }));

  it('Should invoke done callback specified with sync field immediately, and the others after finishing', () =>
    TestPromise(done => {
      result = suite();
      result
        .done('field_1', callback_1)
        .done('field_6', callback_2)
        .done(callback_3);
      expect(callback_1).toHaveBeenCalled();
      expect(callback_2).not.toHaveBeenCalled();
      expect(callback_3).not.toHaveBeenCalled();

      setTimeout(() => {
        expect(callback_2).not.toHaveBeenCalled();
        expect(callback_3).not.toHaveBeenCalled();
        expect(suite.get().hasErrors('field_7')).toBe(true);
        control();
      });

      setTimeout(() => {
        expect(callback_2).toHaveBeenCalled();
        expect(callback_3).toHaveBeenCalled();
        expect(control).toHaveBeenCalled();
        done();
      }, 250);
    }));
});
