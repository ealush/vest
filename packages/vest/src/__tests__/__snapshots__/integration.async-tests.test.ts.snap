// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stateful behavior Should have all fields 1`] = `
Object {
  "field_1": Object {
    "errorCount": 1,
    "errors": Array [
      "field_statement_1",
    ],
    "testCount": 1,
    "valid": false,
    "warnCount": 0,
    "warnings": Array [],
  },
  "field_2": Object {
    "errorCount": 1,
    "errors": Array [
      "field_statement_2",
    ],
    "testCount": 1,
    "valid": false,
    "warnCount": 0,
    "warnings": Array [],
  },
  "field_3": Object {
    "errorCount": 0,
    "errors": Array [],
    "testCount": 1,
    "valid": true,
    "warnCount": 0,
    "warnings": Array [],
  },
  "field_4": Object {
    "errorCount": 0,
    "errors": Array [],
    "testCount": 2,
    "valid": true,
    "warnCount": 1,
    "warnings": Array [
      "field_statement_4",
    ],
  },
  "field_5": Object {
    "errorCount": 2,
    "errors": Array [
      "field_statement_5",
      "field_statement_6",
    ],
    "testCount": 2,
    "valid": false,
    "warnCount": 0,
    "warnings": Array [],
  },
  "field_6": Object {
    "errorCount": 0,
    "errors": Array [],
    "testCount": 1,
    "valid": false,
    "warnCount": 0,
    "warnings": Array [],
  },
  "field_7": Object {
    "errorCount": 0,
    "errors": Array [],
    "testCount": 1,
    "valid": false,
    "warnCount": 0,
    "warnings": Array [],
  },
}
`;
